other/iris/iris.jl

```diff
@@ -1,68 +1,91 @@
-
-
 using Flux
 using Flux: crossentropy, normalise, onecold, onehotbatch
 using Statistics: mean
+using Parameters: @with_kw
 
+@with_kw mutable struct Args
+    lr::Float64 = 0.5
+    repeat::Int = 110
+end
 
-labels = Flux.Data.Iris.labels()
-features = Flux.Data.Iris.features()
-
-
-# Subract mean, divide by std dev for normed mean of 0 and std dev of 1.
-normed_features = normalise(features, dims=2)
-
-
-klasses = sort(unique(labels))
-onehot_labels = onehotbatch(labels, klasses)
-
-
-# Split into training and test sets, 2/3 for training, 1/3 for test.
-train_indices = [1:3:150 ; 2:3:150]
+function get_processed_data(args)
+    labels = Flux.Data.Iris.labels()
+    features = Flux.Data.Iris.features()
 
-X_train = normed_features[:, train_indices]
-y_train = onehot_labels[:, train_indices]
+    # Subract mean, divide by std dev for normed mean of 0 and std dev of 1.
+    normed_features = normalise(features, dims=2)
 
-X_test = normed_features[:, 3:3:150]
-y_test = onehot_labels[:, 3:3:150]
+    klasses = sort(unique(labels))
+    onehot_labels = onehotbatch(labels, klasses)
 
+    # Split into training and test sets, 2/3 for training, 1/3 for test.
+    train_indices = [1:3:150 ; 2:3:150]
 
-# Declare model taking 4 features as inputs and outputting 3 probabiltiies, 
-# one for each species of iris.
-model = Chain(
-    Dense(4, 3),
-    softmax
-)
+    X_train = normed_features[:, train_indices]
+    y_train = onehot_labels[:, train_indices]
 
-loss(x, y) = crossentropy(model(x), y)
+    X_test = normed_features[:, 3:3:150]
+    y_test = onehot_labels[:, 3:3:150]
 
-# Gradient descent optimiser with learning rate 0.5.
-optimiser = Descent(0.5)
+    #repeat the data `args.repeat` times
+    train_data = Iterators.repeated((X_train, y_train), args.repeat)
+    test_data = (X_test,y_test)
 
+    return train_data, test_data
 
-# Create iterator to train model over 110 epochs.
-data_iterator = Iterators.repeated((X_train, y_train), 110)
+end
 
-println("Starting training.")
-Flux.train!(loss, params(model), data_iterator, optimiser)
+# Accuracy Function
+accuracy(x, y, model) = mean(onecold(model(x)) .== onecold(y))
 
-# Evaluate trained model against test set.
-accuracy(x, y) = mean(onecold(model(x)) .== onecold(y))
+# Function to build confusion matrix
+function confusion_matrix(X, y, model)
+    ŷ = onehotbatch(onecold(model(X)), 1:3)
+    y * ŷ
+end
 
-accuracy_score = accuracy(X_test, y_test)
+function train(; kws...)
+	# Initialize hyperparameter arguments
+	args = Args(; kws...)	
+
+	#Loading processed data
+	train_data, test_data = get_processed_data(args)
+
+	# Declare model taking 4 features as inputs and outputting 3 probabiltiies, 
+	# one for each species of iris.
+	model = Chain(
+    		 Dense(4, 3),
+    		 softmax)
+	
+	# Defining loss function to be used in training
+	loss(x, y) = crossentropy(model(x), y)
```

Comment:
for numerical stability use `logitcrossentropy`, remove `softmax` from the model

Reply:
How here using softmax, crossentropy can be numerically unstable?

Reply:
Ok, got it. I have updated that.